atef/cache.py

```diff
@@ -40,12 +65,263 @@ def clear(self) -> None:
         self.pv_to_signal.clear()
 
 
-_signal_cache = None
+@dataclass(frozen=True, eq=True)
+class DataKey:
+    period: Optional[Number] = None
+    method: ReduceMethod = ReduceMethod.average
+    string: bool = False
 
 
-def get_signal_cache() -> _SignalCache[ophyd.EpicsSignalRO]:
-    """Get the global EpicsSignal cache."""
-    global _signal_cache
-    if _signal_cache is None:
-        _signal_cache = _SignalCache(ophyd.EpicsSignalRO)
-    return _signal_cache
+@dataclass(frozen=True, eq=True)
+class ToolKey:
+    tool_cls: Type[tools.Tool]
+    settings: Optional[Hashable]
+
+    @classmethod
+    def from_tool(cls, tool: tools.Tool) -> ToolKey:
+        settings = cast(Hashable, _freeze(dataclasses.asdict(tool)))
+        return cls(
+            tool_cls=type(tool),
+            settings=settings,
+        )
+
+
+def _freeze(data):
```

Comment:
I totally agree. 

On this topic - are there any obvious omissions in this function? I put it together only keeping in mind what was in `Ping`

Reply:
No obvious omissions, but the resulting structure isn't 1:1 interface-compatible with the original for mapping types.

Also, I only just realized now, but the first isinstance check for primitives is seemingly redundant unless you're trying to short-circuit two isinstance calls for primitives.

Reply:
In d6d4e04796e92f817bae67e70f51a5df899d255c:
* I tweaked `Sequence` -> `Iterable` (`set()` is `Iterable` but not a `Sequence`, and `list`/`tuple` are all `Iterable` and `Sequence`)
* I took out the short-circuiting of basic types but only left in `str` because it is iterable (and we don't want to freeze each character)

Reply:
Right, I forgot about str being an iterable- that always bites me randomly